### PR TITLE
refactor: content-manager folder

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* Converts the content-manager folder to TS
	* I think the typings for the recursive data _could_ potentially be better – but we won't know until we use them in the app... recursion is hard
	* The context is looser than I'd like but because we don't have much else typed we don't have the concept of a schema yet, so for now – i think this is good and there's lots of notes for future devs!

### Why is it needed?

* refactor to TS

### Related issue(s)/PR(s)

* closes #17989
* ref https://github.com/strapi/strapi/issues/17690

### CODE REVIEW

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    componentLayoutData: Record<string, unknown>,
+    allComponents: Record<string, unknown>,
+    shouldCheckErrors?: boolean,
+    position?: number
+  ) => void;
+  allLayoutData: {
+    // TODO: this should be a schema?
+    components: Record<string, unknown>;
+    // TODO: this should be a schema?
+    contentType: Record<string, unknown>;
```
Comment (markkaylor) [inline packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:33]: I think the types could be improved by extending [schema](https://github.com/strapi/strapi/blob/main/packages/core/strapi/lib/types/core/schemas/index.d.ts), I see you have the TODOs but maybe we could improve it now, what do you think is best?

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import * as React from 'react';
+
+import { MessageDescriptor } from 'react-intl';
```
Comment (markkaylor) [inline packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:3]: You need this as opposed to the TranslationMessage in the types file?

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  layout: Record<string, unknown>;
+  // TODO: define the tree, can it be inferred?
+  modifiedData: Record<string, unknown>;
+  moveComponentDown?: (dynamicZoneName: string, currentIndex: number) => void;
+  moveComponentField?: (payload: { name: string; newIndex: number; currentIndex: number }) => void;
+  moveComponentUp?: (dynamicZoneName: string, currentIndex: number) => void;
+  onChange?: (
+    payload: { target: { name: string; type: string; value: string } },
+    shouldSetInitialValue?: boolean
+  ) => void;
```
Comment (markkaylor) [inline packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:52]: I'll need this too for the GenericInput could we move it to the types file? I'll share what I had come up with let me know what you think.

```suggestion
  onChange?: (
    payload: { target: { name: string; type?: Attribute.Kind; value: string | number | boolean | Attribute.Media | null } },
    shouldSetInitialValue?: boolean
  ) => void;
```

In my case type is not always passed..which confuses me...

File: packages/core/helper-plugin/src/content-manager/utils/formatContentTypeData.ts [hunk 4]
```diff
@@ -5,8 +5,19 @@ import get from 'lodash/get';
...
-const formatContentTypeData = (data, ct, composSchema) => {
-  const recursiveFormatData = (data, schema) => {
+const formatContentTypeData = <
+  TData extends Record<string, unknown>,
+  TSchema extends Record<string, unknown>,
+  TComponentSchema extends Record<string, object>
+>(
+  data: TData,
+  ct: TSchema,
+  composSchema: TComponentSchema
```
Comment (markkaylor) [inline packages/core/helper-plugin/src/content-manager/utils/formatContentTypeData.ts:15]: Not sure exactly what these are but maybe they could be typed with the strapi [schema](https://github.com/strapi/strapi/blob/main/packages/core/strapi/lib/types/core/schemas/index.d.ts)?

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx (hunk 1 above)
Comment (joshuaellis) [reply packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:33]: aha, i didn't realise we had them, i'll do some investigation to ensure they're _exactly_ what we want, but nice catch ✨

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx (hunk 2 above)
Comment (joshuaellis) [reply packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:3]: Looking at how `formErrors` was used, it looked like it was only ever used with the `MessageDescriptor` shape. I think we can change it when working more in the CM if it doesn't fit in reality?

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx (hunk 3 above)
Comment (joshuaellis) [reply packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:52]: I would move it when you're doing GenericInput to see how it works with that and see if you actually need it shared as opposed to anticipating, just incase for whatever reason, you don't need it and it's automatically inferred.

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  isSingleType: boolean;
+  layout?: Schema.CollectionType | Schema.SingleType;
+  // TODO: define the tree, can it be inferred?
+  modifiedData: Record<string, unknown>;
+  moveComponentDown?: (dynamicZoneName: string, currentIndex: number) => void;
+  moveComponentField?: (payload: { name: string; newIndex: number; currentIndex: number }) => void;
+  moveComponentUp?: (dynamicZoneName: string, currentIndex: number) => void;
+  onChange?: <TAttribute extends Attribute.Any>(
+    payload: {
+      target: { name: string; type?: TAttribute['type']; value: Attribute.GetValue<TAttribute> };
```
Comment (markkaylor) [inline packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:50]: Sorry sent you in the wrong direction here as I mentioned on discord, type is an input type and not a strapi type. I've got something in place over on https://github.com/strapi/strapi/pull/18045/files that I think maybe makes some sense. If this works for now that's fine we can maybe figure out the difference between the two and what's needed for each after merging.

**Review by markkaylor — APPROVED** [review]

File: packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx (hunk 5 above)
Comment (joshuaellis) [reply packages/core/helper-plugin/src/content-manager/CMEditViewDataManager.tsx:50]: Lets merge this, I'll take a look at your PR and do the changes necessary there 👀

**Pull request merged**